lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
          * If user declares "foo", convert to "foo:false".
          */
         this.globals = (options.globals || []).reduce((globals, def) => {
-            const parts = def.split(":");
+            const parts = def.split(SUBCONFIG_SEP);
 
             globals[parts[0]] = (parts.length > 1 && parts[1] === "true");
 
             return globals;
         }, {});
 
-        const useConfig = options.configFile;
-
-        this.options = options;
+        let useConfig = options.configFile;
 
         if (useConfig) {
             debug(`Using command line config ${useConfig}`);
-            if (isResolvable(useConfig) || isResolvable(`eslint-config-${useConfig}`) || useConfig.charAt(0) === "@") {
-                this.useSpecificConfig = loadConfig(useConfig);
-            } else {
-                this.useSpecificConfig = loadConfig(path.resolve(this.options.cwd, useConfig));
+            if (!(isResolvable(useConfig) || isResolvable(`eslint-config-${useConfig}`) || useConfig.charAt(0) === "@")) {
+                useConfig = path.resolve(this.options.cwd, useConfig);
             }
+            this.useSpecificConfig = ConfigFile.loadCached(useConfig);
+        }
+
+        if (this.options.plugins) {
+            Plugins.loadAll(this.options.plugins);
         }
+
+        // Empty values in configs don't merge properly
+        const cliConfigOptions = {
+            env: this.env,
+            rules: this.options.rules,
+            globals: this.globals,
+            parserOptions: this.parserOptions,
+            plugins: this.options.plugins
+        };
+
+        this.cliConfig = {};
+        Object.keys(cliConfigOptions).forEach(configKey => {
+            const value = cliConfigOptions[configKey];
+
+            if (value) {
+                this.cliConfig[configKey] = value;
+            }
+        });
```

Comment:
See above. Passing `plugins` as `undefined` threw a hard error, and setting other keys to `undefined` made tests fail.